txn/meta.go

```diff
@@ -0,0 +1,99 @@
+// Copyright (C) MongoDB, Inc. 2019-present.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License. You may obtain
+// a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+
+package txn
+
+import (
+	"github.com/mongodb/mongo-tools-common/db"
+)
+
+// "empty" prevOpTime is {ts: Timestamp(0, 0), t: NumberLong(-1)} as BSON.
+var emptyPrev = string([]byte{
+	28, 0, 0, 0, 17, 116, 115, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+	18, 116, 0, 255, 255, 255, 255, 255, 255, 255, 255, 0,
+})
+
+// ID wraps fields needed to uniquely identify a transaction for use as a map
+// key.  The 'lsid' is a string rather than bson.Raw or []byte so that this
+// type is a valid map key.
+type ID struct {
+	lsid      string
+	txnNumber int64
+}
+
+// Meta holds information extracted from an oplog entry for later routing
+// logic.  Zero value means 'not a transaction'.  We store 'prevOpTime' as
+// string so the struct is comparable.
+type Meta struct {
+	id         ID
+	commit     bool
+	abort      bool
+	partial    bool
+	prepare    bool
+	prevOpTime string
+}
+
+// NewMeta extracts transaction metadata from an oplog entry.  A non-transaction will
+// return a zero-value Meta struct, not an error.
+func NewMeta(op db.Oplog) (Meta, error) {
```

Comment:
I don't see any cases where this method actually errors. do you see that changing in the future?

Reply:
Not currently, but in the future I'd like to change `Object` to be `bson.Raw` so it's not unmarshaling the entire, deep data structure of documents within the applyOps.  Then working with a raw value could potentially error, even if unlikely.  I'll add a comment about it.